lib/Frontend/Frontend.cpp

```diff
@@ -497,58 +565,131 @@ void CompilerInstance::performSema() {
   if (options.DebugTimeExpressionTypeChecking) {
     TypeCheckOptions |= TypeCheckingFlags::DebugTimeExpressions;
   }
+  return TypeCheckOptions;
+}
+
+// Return true if had load error
+bool CompilerInstance::parsePartialModulesAndLibraryFiles(
+    SourceFile::ImplicitModuleImportKind implicitModuleImportKind,
+    ModuleDecl *objCModuleUnderlyingMixedFramework,
+    ModuleDecl *importedHeaderModule,
+    SmallVectorImpl<ModuleDecl *> &importModules,
+    PersistentParserState &PersistentState,
+    DelayedParsingCallbacks *DelayedParseCB) {
+  SharedTimer timer("performSema-parsePartialModulesAndLibraryFiles");
+  bool hadLoadError = false;
+  // Parse all the partial modules first.
+  for (auto &PM : PartialModules) {
+    assert(PM.ModuleBuffer);
+    if (!SML->loadAST(*MainModule, SourceLoc(), std::move(PM.ModuleBuffer),
+                      std::move(PM.ModuleDocBuffer)))
+      hadLoadError = true;
+  }
+
+  // Then parse all the library files.
+  for (auto BufferID : BufferIDs) {
+    if (BufferID != MainBufferID)
+      parseALibraryFile(BufferID, implicitModuleImportKind,
+                        objCModuleUnderlyingMixedFramework,
+                        importedHeaderModule, importModules, PersistentState,
+                        DelayedParseCB);
+  }
+  if (Invocation.isCodeCompletion()) {
+    // When we are doing code completion, make sure to emit at least one
+    // diagnostic, so that ASTContext is marked as erroneous.  In this case
+    // various parts of the compiler (for example, AST verifier) have less
+    // strict assumptions about the AST.
+    Diagnostics.diagnose(SourceLoc(), diag::error_doing_code_completion);
```

Comment:
This doesn't really have anything to do with parsing *these* files, though.

Reply:
Here is the code before I touched it:

 if (Invocation.isCodeCompletion()) {
    // When we are doing code completion, make sure to emit at least one
    // diagnostic, so that ASTContext is marked as erroneous.  In this case
    // various parts of the compiler (for example, AST verifier) have less
    // strict assumptions about the AST.
    Diagnostics.diagnose(SourceLoc(), diag::error_doing_code_completion);
  }

What would you like me to do with it? Where should the comment go?

Reply:
The code still has to happen, but it isn't part of parsing. It's not really part of *any* of the other high-level operations, but it has to happen before type-checking because AST verification happens at the end of type-checking.

Reply:
I guess I would just lift it up to the immediate caller.

Reply:
OK, done. It seems weird the the diagnostic is issued even if there was a load error, but that is what the behavior was before, so I've preserved it.